src/main/java/fi/otavanopisto/kuntaapi/server/rest/ServicesApiImpl.java

```diff
@@ -0,0 +1,510 @@
+    
+    for (ServiceChannelProvider serviceChannelProvider : getServiceChannelProviders()) {
+      result.addAll(serviceChannelProvider.listPhoneChannels(serviceId));
+    }
+    
+    int resultCount = result.size();
+    int firstIndex = firstResult == null ? 0 : Math.min(firstResult.intValue(), resultCount);
+    int toIndex = maxResults == null ? resultCount : Math.min(firstIndex + maxResults.intValue(), resultCount);
+    
+    return Response.ok(result.subList(firstIndex, toIndex))
+      .build();
+  }
+
+  @Override
+  public Response listServicePrintableFormChannels(String serviceIdParam, Long firstResult, Long maxResults) {
+    ServiceId serviceId = toServiceId(serviceIdParam);
+    if (serviceId == null) {
+      return createBadRequest(String.format(INVALID_SERVICE_ID, serviceIdParam));
+    }
+    
+    if (firstResult != null && firstResult < 0) {
+      return createBadRequest(FIRST_RESULT_MUST_BY_A_POSITIVE_INTEGER);
+    }
+    
+    if (maxResults != null && maxResults < 0) {
+      return createBadRequest(MAX_RESULTS_MUST_BY_A_POSITIVE_INTEGER);
+    }
+    
+    List<PrintableFormChannel> result = new ArrayList<>();
+    
+    for (ServiceChannelProvider serviceChannelProvider : getServiceChannelProviders()) {
+      result.addAll(serviceChannelProvider.listPrintableFormChannels(serviceId));
+    }
+    
+    int resultCount = result.size();
+    int firstIndex = firstResult == null ? 0 : Math.min(firstResult.intValue(), resultCount);
+    int toIndex = maxResults == null ? resultCount : Math.min(firstIndex + maxResults.intValue(), resultCount);
+    
+    return Response.ok(result.subList(firstIndex, toIndex))
+      .build();
+  }
+
+  @Override
+  public Response listServiceServiceLocationChannels(String serviceIdParam, Long firstResult, Long maxResults) {
+    ServiceId serviceId = toServiceId(serviceIdParam);
+    if (serviceId == null) {
+      return createBadRequest(String.format(INVALID_SERVICE_ID, serviceIdParam));
+    }
+    
+    if (firstResult != null && firstResult < 0) {
+      return createBadRequest(FIRST_RESULT_MUST_BY_A_POSITIVE_INTEGER);
+    }
+    
+    if (maxResults != null && maxResults < 0) {
+      return createBadRequest(MAX_RESULTS_MUST_BY_A_POSITIVE_INTEGER);
+    }
+    
+    List<ServiceLocationChannel> result = new ArrayList<>();
+    
+    for (ServiceChannelProvider serviceChannelProvider : getServiceChannelProviders()) {
+      result.addAll(serviceChannelProvider.listServiceLocationChannels(serviceId));
+    }
+    
+    int resultCount = result.size();
+    int firstIndex = firstResult == null ? 0 : Math.min(firstResult.intValue(), resultCount);
+    int toIndex = maxResults == null ? resultCount : Math.min(firstIndex + maxResults.intValue(), resultCount);
+    
+    return Response.ok(result.subList(firstIndex, toIndex))
+      .build();
+  }
+
+  @Override
+  public Response listServiceWebPageChannels(String serviceIdParam, Long firstResult, Long maxResults) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) The Cyclomatic Complexity of this method "listServiceWebPageChannels" is 12 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)